granulate_utils/metadata/cloud.py

```diff
@@ -95,6 +102,22 @@ def get_aws_metadata() -> Optional[AwsInstanceMetadata]:
     )
 
 
+def get_aws_container_metadata() -> Optional[AwsContainerMetadata]:
+    ecs_container_metadata_uri_v4 = os.environ.get("ECS_CONTAINER_METADATA_URI_V4")
+    if ecs_container_metadata_uri_v4 is None:
+        return None
+    response = send_request(ecs_container_metadata_uri_v4)
+    if response is None:
+        return None
+    metadata = response.json()
+    return AwsContainerMetadata(
+        provider="aws",
+        execution_env=os.environ["AWS_EXECUTION_ENV"],
+        region=os.environ["AWS_REGION"],
```

Comment:
Make it optional then?

Reply:
the existence of this envar is optional, but we always need to know the region

Reply:
empirically it exists on Fargate, which is what we are currently concerned with. are we concerned with non-Fargate environments?

Reply:
We are concerned with region, but we get it from the instance metadata, i.e `get_aws_instance_metadata`.

This raises a problem with your design - `get_static_cloud_metadata` returns the first match, and only it, because it segregates based on providers. What you're adding here is additional types of metadata that we want to be collected together & merged & presented as a single result to the backend.

Practically speaking, I suggest we currently add the `execution_env` in ALL cases to the AWS metadata we return. And specifically for fargate, we will send the request to ECS_CONTAINER_METADATA_URI_V4 to get the container ARN, and add it to the metadata as well. In other cases (all non-fargate) it can be none.
Please check if instance metadata is available in fargate - if it does then we do not need to concern ourselves with getting the region from additional sources.

Reply:
instance metadata is not available in fargate, because tasks are not running on an instance. they run in a dedicated vm

Reply:
So in fargate - the instance metadata fetcher will fail, but in ECS, but fetchers will succeed and arbitrarily one of them will be returned. Therefore please merge them to a single one as I suggested.

Reply:
what do you mean arbitrarily?

Reply:
> get_static_cloud_metadata returns the first match, and only it, because it segregates based on providers

It calls all fetchers in parallel and returns the first one that has a result.

Reply:
yes. does gcp or azure metadata work on ecs?

Reply:
Discussed f2f - `get_fargate_metadata`

Reply:
@Jongy done